test/jdk/jdk/internal/ref/Cleaner/CleanableListTest.java

```diff
@@ -0,0 +1,126 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8343704
+ * @library /test/lib
+ * @compile/module=java.base jdk/internal/ref/CleanableListTestHelper.java jdk/internal/ref/TestCleanable.java
+ * @modules java.base/jdk.internal.ref
+ * @run testng/othervm CleanableListTest
+ */
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Random;
+import jdk.internal.ref.CleanableListTestHelper;
+import jdk.internal.ref.TestCleanable;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.testng.annotations.Ignore;
+
+public class CleanableListTest {
+
+    static final int SINGLE_NODE_CAPACITY = CleanableListTestHelper.NODE_CAPACITY;
+    static final int MULTI_NODE_CAPACITY = CleanableListTestHelper.NODE_CAPACITY * 4;
+
+    static final int RANDOM_ITERATIONS = 10_000_000;
+
+    @Test
+    public void testSingle() {
+        CleanableListTestHelper list = new CleanableListTestHelper();
+        Assert.assertTrue(list.isEmpty());
+        TestCleanable tc = list.newCleanable();
+        Assert.assertFalse(list.isEmpty());
+        Assert.assertTrue(list.remove(tc));
+        Assert.assertTrue(list.isEmpty());
+        Assert.assertFalse(list.remove(tc));
+    }
+
+    @Test
+    public void testSequential_Single() {
+        doSequential(SINGLE_NODE_CAPACITY);
+    }
+
+    @Test
+    public void testSequential_Multi() {
+        doSequential(MULTI_NODE_CAPACITY);
+    }
+
+    private void doSequential(int size) {
+        CleanableListTestHelper list = new CleanableListTestHelper();
+        Assert.assertTrue(list.isEmpty());
+
+        List<TestCleanable> tcs = new ArrayList<>();
+        for (int c = 0; c < size; c++) {
+            tcs.add(list.newCleanable());
+        }
+        Assert.assertFalse(list.isEmpty());
+
+        for (TestCleanable tc : tcs) {
+            Assert.assertTrue(list.remove(tc));
+        }
+        Assert.assertTrue(list.isEmpty());
+    }
+
+    @Test
+    public void testRandom_Single() {
+        doRandom(SINGLE_NODE_CAPACITY);
+    }
+
+    @Test
+    public void testRandom_Multi() {
+        doRandom(MULTI_NODE_CAPACITY);
+    }
+
+    private void doRandom(int size) {
+        CleanableListTestHelper list = new CleanableListTestHelper();
+        Assert.assertTrue(list.isEmpty());
+
+        Random r = new Random(42);
```

Comment:
If a constant random seed is used, the test will always run in the same way, yes?
Also, consider using `jdk.test.lib.RandomFactory`.

Reply:
Migrated to `RandomFactory` now.